ecs-agent/acs/session/session_test.go

```diff
@@ -0,0 +1,1328 @@
+			Return(mockWsClient),
+		mockWsClient.EXPECT().Connect(gomock.Any(), gomock.Any(), gomock.Any()).Do(func(interface{},
+			interface{}, interface{}) {
+			cancel()
+		}).Return(time.NewTimer(wsclient.DisconnectTimeout), nil),
+	)
+	acsSession := NewSession(testconst.ContainerInstanceARN,
+		testconst.ClusterName,
+		discoverEndpointClient,
+		nil,
+		noopFunc,
+		mockClientFactory,
+		metricsfactory.NewNopEntryFactory(),
+		agentVersion,
+		agentGitShortHash,
+		dockerVersion,
+		nil,
+		nil,
+		nil,
+		nil,
+		nil,
+		nil,
+		nil,
+		nil,
+		nil,
+		nil,
+		nil,
+		nil,
+	)
+	acsSession.(*session).backoff = mockBackoff
+	acsSession.(*session).heartbeatTimeout = 20 * time.Millisecond
+	acsSession.(*session).heartbeatJitter = 10 * time.Millisecond
+	acsSession.(*session).disconnectTimeout = 30 * time.Millisecond
+	acsSession.(*session).disconnectJitter = 10 * time.Millisecond
+
+	err := acsSession.Start(ctx)
+	assert.Equal(t, context.Canceled, err)
+}
+
+// TestStartSessionHandlesAttachResourceMessages tests that the Session is able to handle attach
+// resource messages when the Session's resourceHandler is not nil and its dockerVersion is "containerd".
+func TestStartSessionHandlesAttachResourceMessages(t *testing.T) {
+	ctrl := gomock.NewController(t)
+	defer ctrl.Finish()
+
+	resourceHandler := mock_session.NewMockResourceHandler(ctrl)
+	discoverEndpointClient := mock_api.NewMockECSDiscoverEndpointSDK(ctrl)
+	ctx, cancel := context.WithCancel(context.Background())
+	closeWS := make(chan bool)
+	fakeServer, serverIn, requestsChan, errChan, err := startFakeACSServer(closeWS)
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer close(serverIn)
+
+	go func() {
+		for {
+			select {
+			case <-requestsChan:
+				// Cancel the context when we get the ACK request.
+				cancel()
+			}
+		}
+	}()
+
+	// DiscoverPollEndpoint returns the URL for the server that we started.
+	discoverEndpointClient.EXPECT().DiscoverPollEndpoint(testconst.ContainerInstanceARN).Return(fakeServer.URL, nil)
+
+	ended := make(chan bool, 1)
+	go func() {
+		acsSession := NewSession(testconst.ContainerInstanceARN,
+			testconst.ClusterName,
+			discoverEndpointClient,
+			testCreds,
+			noopFunc,
+			acsclient.NewACSClientFactory(),
+			metricsfactory.NewNopEntryFactory(),
+			agentVersion,
+			agentGitShortHash,
+			"containerd",
+			testMinAgentConfig,
+			nil,
+			nil,
+			nil,
+			nil,
+			nil,
+			nil,
+			nil,
+			nil,
+			nil,
+			resourceHandler,
+			nil,
+		)
+
+		acsSession.Start(ctx)
+		// Start should never return unless the context is canceled.
+		ended <- true
+	}()
+
+	// WaitGroup is necessary to wait for HandleResourceAttachment to be called in separate goroutine before exiting
+	// the test.
+	wg := sync.WaitGroup{}
+	wg.Add(1)
+	resourceHandler.EXPECT().HandleResourceAttachment(gomock.Any()).Do(func(arg0 interface{}) {
+		defer wg.Done() // decrement WaitGroup counter now that HandleResourceAttachment function has been called
```

Comment:
nit: comment capitalization and punctutation

Reply:
I typically don't capitalize or add period at the end if a code comment is in-line with a line of code (i.e., not on its own separate line), as I consider it akin to more of a "bullet point". If you have a strong preference for this though, I don't mind updating it to match your preference.

Reply:
no strong preference here either, just wanted to call it out